nav2_system_tests/src/costmap_filters/filters_tester.cpp

```diff
@@ -0,0 +1,334 @@
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  is_active_ = false;
+
+  transform_timer_.reset();
+  tf_broadcaster_.reset();
+
+  costmap_ros_->on_deactivate(state);
+  planner_->deactivate();
+
+  // destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FiltersTester::on_cleanup(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  costmap_ros_->on_cleanup(state);
+  planner_->cleanup();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FiltersTester::on_error(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FiltersTester::on_shutdown(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  costmap_ros_->on_shutdown(state);
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void FiltersTester::startRobotTransform()
+{
+  // Provide the robot pose transform
+  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
+
+  // Set an initial pose
+  geometry_msgs::msg::Point robot_position;
+  robot_position.x = 0.0;
+  robot_position.y = 0.0;
+  updateRobotPosition(robot_position);
+
+  // Publish the transform periodically
+  transform_timer_ = create_wall_timer(
+    100ms, std::bind(&FiltersTester::publishRobotTransform, this));
+}
+
+void FiltersTester::updateRobotPosition(const geometry_msgs::msg::Point & position)
+{
+  if (!base_transform_) {
+    base_transform_ = std::make_unique<geometry_msgs::msg::TransformStamped>();
+    base_transform_->header.frame_id = "map";
+    base_transform_->child_frame_id = "base_link";
+  }
+
+  base_transform_->header.stamp = now() + rclcpp::Duration(250000000);
+  base_transform_->transform.translation.x = position.x;
+  base_transform_->transform.translation.y = position.y;
+  base_transform_->transform.rotation.w = 1.0;
+
+  publishRobotTransform();
+}
+
+void FiltersTester::publishRobotTransform()
+{
+  if (base_transform_) {
+    tf_broadcaster_->sendTransform(*base_transform_);
+  }
+}
+
+void FiltersTester::spinTester()
+{
+  if (rclcpp::ok()) {
+    // Spin LifycycleNode and Costmap2DROS
+    rclcpp::spin_some(this->get_node_base_interface());
+    rclcpp::spin_some(costmap_ros_->get_node_base_interface());
+  }
+}
+
+bool FiltersTester::testPlan(
+  const geometry_msgs::msg::PoseStamped & start,
+  const geometry_msgs::msg::PoseStamped & end)
+{
+  geometry_msgs::msg::PoseStamped pose = start;
+  nav_msgs::msg::Path path;
+  bool result;
+
+  do {
+    spinTester();
+
+    result = checkPlan(pose, end, path);
+    if (result) {
+      if (path.poses.size() > 2) {
+        // Goal is not reached, continue moving
+        // printPath(path);
+        pose = path.poses[1];
+        if (isInKeepout(pose.pose.position)) {
```

Comment:
Why aren't you testing for all points in the path? Why is there a while loop? You should only need to plan once to know if its working or not.

Reply:
If this test also had another keepout layer in the local costmap in a different position than the keep-out marked zone in the global, you could also track the robot pose and make sure the local rolling is working too.

Reply:
> Why aren't you testing for all points in the path? Why is there a while loop? You should only need to plan once to know if its working or not.

That is because at the initial moment path planner might not know about whole mask on global costmap (`min_i, min_j, max_i, max_j` window did not reached keepout zone yet. This window is being restricted by obstacle layer without having any observation sources like laser scan, which is not provided by test). Therefore, initially planner tries to produce the path through keepout zones at the moment without knowledge about keepout zone (which will appear later after window bounds will reach keepout zone). So, that is why we need to have a step-by-step robot moving using this while-loop.

Reply:
> If this test also had another keepout layer in the local costmap in a different position than the keep-out marked zone in the global, you could also track the robot pose and make sure the local rolling is working too.

I think, this also better to cover by real-time python system test, not in here. This is because, this test is using only global costmap and global path planner.

Reply:
> This window is being restricted by obstacle layer without having any observation sources like laser scan, which is not provided by test).

The point of this test isn't in testing the perception system, its in testing the keepout layer. Why not include a static layer so that the size is already set as full and remove the obstacle layer? That way you have full visibility. The first update the static map will have the full map so then the keep out will cover the full min/max size. Then you only need to plan once and check the full path.

Reply:
I've changed layers and their order from static->obstacle->keepout` to `static->keepout->inflation` and made test to be static instead of step-by-step in https://github.com/ros-planning/navigation2/commit/ab3f5fe2e5f47906832bfee91c2b6ff8e280eb8d#diff-606e794691abfc30db44db2a48a199f6

Reply:
I think it shoudl be static->inflation->keepout so that the keepout layer itself isn't inflated?

Reply:
I've initially thought to make keepout after inflation in order to avoid path planner to make the path on keepout's boundary. However, after the test was checked in 10~20 minutes of loop-run, it turned out that it works also with `static->inflation->keepout` order. And, yes, in all examples keepout filter also goes after all layers applied. So, this was changed to correct order in the last patch.

Reply:
Also, I've added some workaround here: https://github.com/ros-planning/navigation2/pull/1882/commits/a8fb4ed6703c56e10def3393ffa6cd96d8fdd8fc#diff-606e794691abfc30db44db2a48a199f6R234 in order to avoid some rare fails appear when `filter_mask` comes to `keepout_filter` later after `static_layer` has been worked-off.